### PR TITLE
Add SQL query planner support for `DISTRIBUTE BY`

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3207

 # Rationale for this change

Support Dask SQL queries that use this syntax

# What changes are included in this PR?

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -973,6 +973,20 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        }
+        if !select.lateral_views.is_empty() {
+            return Err(DataFusionError::NotImplemented("LATERAL VIEWS".to_string()));
+        }
+        if select.qualify.is_some() {
+            return Err(DataFusionError::NotImplemented("QUALIFY".to_string()));
+        }
+        if select.top.is_some() {
+            return Err(DataFusionError::NotImplemented("TOP".to_string()));
+        }
```
Comment (andygrove) [inline datafusion/sql/src/planner.rs:988]: We were previously just ignoring these fields so I think it is better to error if they are set and we don't support them

**Comment by andygrove** [comment]
@jdye64 Do you think this is sufficient?

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -1390,8 +1399,9 @@ pub enum Partitioning {
     RoundRobinBatch(usize),
     /// Allocate rows based on a hash of one of more expressions and the specified number
     /// of partitions.
-    /// This partitioning scheme is not yet fully supported. See <https://issues.apache.org/jira/browse/ARROW-11011>
```
Comment (andygrove) [inline datafusion/expr/src/logical_plan/plan.rs:1393]: This issue (https://github.com/apache/arrow-datafusion/issues/131) has been closed so I removed the comment

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -1390,8 +1399,9 @@ pub enum Partitioning {
...
     Hash(Vec<Expr>, usize),
+    /// The DISTRIBUTE BY clause is used to repartition the data based on the input expressions
+    DistributeBy(Vec<Expr>),
```
Comment (andygrove) [inline datafusion/expr/src/logical_plan/plan.rs:1404]: Another option here would be to re-use `Hash` and make the number of partitions an `Option<usize>`.

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:1404]: Given the description in the spark docs https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-distribute-by.html I think either hashing or not would produce the same behavior with respect to partitioning.

However, using just the expressions as you have in this PR is probably more performant -- hashing the values would be wasted work.

File: datafusion/core/src/physical_plan/planner.rs [hunk 4]
```diff
@@ -729,6 +729,9 @@ impl DefaultPhysicalPlanner {
                                 .collect::<Result<Vec<_>>>()?;
                             Partitioning::Hash(runtime_expr, *n)
                         }
+                        LogicalPartitioning::DistributeBy(_) => {
+                            return Err(DataFusionError::NotImplemented("Physical plan does not support DistributeBy partitioning".to_string()))
```
Comment (alamb) [inline datafusion/core/src/physical_plan/planner.rs:733]: Thanks for breaking up these projects into smaller, easier to review chunks 👍

**Review by alamb — APPROVED** [review]
LGTM

**Pull request merged**